lang/funcs/collection.go

```diff
@@ -346,12 +346,14 @@ var MapFunc = function.New(&function.Spec{
 
 		for i := 0; i < len(args); i += 2 {
 
-			key := args[i].AsString()
-
-			err := gocty.FromCtyValue(args[i], &key)
+			keyVal, err := convert.Convert(args[i], cty.String)
 			if err != nil {
 				return cty.NilVal, err
 			}
+			if keyVal.IsNull() {
+				return cty.NilVal, fmt.Errorf("argument %d is a null key", i+1)
```

Comment:
Not super important here, but just wanted to point out that you can return the special error type [`ArgError`](https://pkg.go.dev/github.com/zclconf/go-cty@v1.3.1/cty/function?tab=doc#ArgError) from a function (e.g. using `NewArgErrorf` instead of `Errorf` here) to indicate that a specific argument is incorrect.

When using functions with HCL, HCL recognizes that error type and indicates the specific single argument as the problem, rather than the entire call.